File: electron/main/managed/git/mirror.ts
import { lstat, rename, stat } from 'node:fs/promises'
import { randomUUID } from 'node:crypto'
import nodePath from 'node:path'
import { requireSingleGitLine, runRequiredGitCommand } from './command'
import { GitRuntimeError, gitRuntimeFailure } from './errors'
import {
  acquireManagedGitOperationLock,
  assertManagedGitTarget,
  ensureManagedGitInstallationDirectories
} from './paths'
import {
  assertGitNamedRemote,
  assertGitRemoteIdentity,
  assertGitRemoteName,
  parseGitRemoteSource
} from './remote'
import { requireGitCommandSuccess, type GitCommandRunner } from './process'
import type {
  GitExecutableRegistration,
  GitExecutionContext,
  GitNamedRemote,
  ManagedGitInstallationPaths
} from './types'

/** Readback identity for a verified launcher-owned bare mirror. */
export interface ManagedGitMirror {
  readonly path: string
  readonly remote: GitNamedRemote
}

/** Creates a new bare mirror through a private staging path and one atomic publication rename. */
export async function createManagedGitMirror(
  runner: GitCommandRunner,
  registration: GitExecutableRegistration,
  context: GitExecutionContext,
  paths: ManagedGitInstallationPaths,
  remote: GitNamedRemote
): Promise<ManagedGitMirror> {
  assertNamedRemoteIntegrity(remote)
  const lock = await acquireManagedGitOperationLock(paths)
  try {
    await ensureManagedGitInstallationDirectories(paths)
    await assertPathAbsent(
      paths.mirrorPath,
      'git.mirror_exists',
      'Managed Git mirror already exists.'
    )
    const stagingMirrorPath = nodePath.join(paths.stagingPath, `mirror-${randomUUID()}.git`)
    await assertPathAbsent(
      stagingMirrorPath,
      'git.operation_failed',
      'Managed Git staging mirror already exists.'
    )
    await assertSameFilesystem(
      paths.stagingPath,
      nodePath.dirname(paths.mirrorPath),
      paths.installationId
    )

    await runRequiredGitCommand(runner, registration, context, {
      operation: 'git.create_bare_mirror',
      arguments: ['init', '--bare', stagingMirrorPath]
    })
    await runRequiredGitCommand(runner, registration, context, {
      operation: 'git.add_mirror_remote',
      arguments: [
        '--git-dir',
        stagingMirrorPath,
        'remote',
        'add',
        remote.name,
        remote.source.declaredUrl
      ]
    })
    await fetchMirrorRemote(runner, registration, context, stagingMirrorPath, remote.name)
    await verifyBareMirror(runner, registration, context, stagingMirrorPath, remote)
    try {
      await rename(stagingMirrorPath, paths.mirrorPath)
    } catch (error) {
      throw gitRuntimeFailure(
        'git.operation_failed',
        'Managed Git mirror could not be published atomically.',
        error,
        {
          installationId: paths.installationId
        }
      )
    }
    await assertManagedGitTarget(paths, paths.mirrorPath, 'Managed Git mirror')
    await verifyBareMirror(runner, registration, context, paths.mirrorPath, remote)
    return { path: paths.mirrorPath, remote }
  } finally {
    await lock.release()
  }
}

/**
 * Imports one previously verified bundle into a fresh managed mirror without persisting the bundle as a Git remote.
 *
 * Branches are written directly to the normal remote-tracking namespace so subsequent revision resolution is
 * identical to a mirror created from the declared network remote.
 */
export async function createManagedGitMirrorFromBundle(
  runner: GitCommandRunner,
  registration: GitExecutableRegistration,
  context: GitExecutionContext,
  paths: ManagedGitInstallationPaths,
  remote: GitNamedRemote,
  bundlePath: string
): Promise<ManagedGitMirror> {
  assertNamedRemoteIntegrity(remote)
  await assertDirectRegularBundleFile(bundlePath)
  const lock = await acquireManagedGitOperationLock(paths)
  try {
    await ensureManagedGitInstallationDirectories(paths)
    await assertPathAbsent(
      paths.mirrorPath,
      'git.mirror_exists',
      'Managed Git mirror already exists.'
    )
    const stagingMirrorPath = nodePath.join(paths.stagingPath, `mirror-${randomUUID()}.git`)
    await assertPathAbsent(
      stagingMirrorPath,
      'git.operation_failed',
      'Managed Git staging mirror already exists.'
    )
    await assertSameFilesystem(
      paths.stagingPath,
      nodePath.dirname(paths.mirrorPath),
      paths.installationId
    )

    await runRequiredGitCommand(runner, registration, context, {
      operation: 'git.create_bare_mirror',
      arguments: ['init', '--bare', stagingMirrorPath]
    })
    await runRequiredGitCommand(runner, registration, context, {
      operation: 'git.import_bundle_objects',
      arguments: [
        '--git-dir',
        stagingMirrorPath,
        'fetch',
        '--atomic',
        '--no-write-fetch-head',
        '--no-tags',
        '--',
        bundlePath,
        `+refs/heads/*:refs/remotes/${remote.name}/*`,
        '+refs/tags/*:refs/tags/*'
      ]
    })
    await runRequiredGitCommand(runner, registration, context, {
      operation: 'git.add_mirror_remote',
      arguments: [
        '--git-dir',
        stagingMirrorPath,
        'remote',
        'add',
        remote.name,
        remote.source.declaredUrl
      ]
    })
    await verifyBareMirror(runner, registration, context, stagingMirrorPath, remote)
    try {
      await rename(stagingMirrorPath, paths.mirrorPath)
    } catch (error) {
      throw gitRuntimeFailure(
        'git.operation_failed',
        'Managed Git mirror could not be published atomically.',
        error,
        {
          installationId: paths.installationId
        }
      )
    }
    await assertManagedGitTarget(paths, paths.mirrorPath, 'Managed Git mirror')
    await verifyBareMirror(runner, registration, context, paths.mirrorPath, remote)
    return { path: paths.mirrorPath, remote }
  } finally {
    await lock.release()
  }
}

/** Fetches only the declared remote into an existing verified launcher-owned mirror. */
export async function fetchManagedGitMirror(
  runner: GitCommandRunner,
  registration: GitExecutableRegistration,
  context: GitExecutionContext,
  paths: ManagedGitInstallationPaths,
  remote: GitNamedRemote
): Promise<ManagedGitMirror> {
  assertNamedRemoteIntegrity(remote)
  const lock = await acquireManagedGitOperationLock(paths)
  try {
    await assertManagedGitTarget(paths, paths.mirrorPath, 'Managed Git mirror')
    await verifyBareMirror(runner, registration, context, paths.mirrorPath, remote)
    await fetchMirrorRemote(runner, registration, context, paths.mirrorPath, remote.name)
    await verifyBareMirror(runner, registration, context, paths.mirrorPath, remote)
    return { path: paths.mirrorPath, remote }
  } finally {
    await lock.release()
  }
}

/** Reads a managed mirror without fetch or configuration mutation. */
export async function inspectManagedGitMirror(
  runner: GitCommandRunner,
  registration: GitExecutableRegistration,
  context: GitExecutionContext,
  paths: ManagedGitInstallationPaths,
  remote: GitNamedRemote
): Promise<ManagedGitMirror> {
  assertNamedRemoteIntegrity(remote)
  await assertManagedGitTarget(paths, paths.mirrorPath, 'Managed Git mirror')
  await verifyBareMirror(runner, registration, context, paths.mirrorPath, remote)
  return { path: paths.mirrorPath, remote }
}

/** Verifies both bare-repository state and the single expected remote identity. */
export async function verifyBareMirror(
  runner: GitCommandRunner,
  registration: GitExecutableRegistration,
  context: GitExecutionContext,
  mirrorPath: string,
  remote: GitNamedRemote
): Promise<void> {
  const bare = await runRequiredGitCommand(runner, registration, context, {
    operation: 'git.verify_bare_mirror',
    arguments: ['--git-dir', mirrorPath, 'rev-parse', '--is-bare-repository']
  })
  if (requireSingleGitLine(bare.stdout, 'Bare mirror state') !== 'true') {
    throw new GitRuntimeError(
      'git.repository_not_bare',
      'Managed Git mirror is not a bare repository.'
    )
  }
  const [remoteNames, remoteUrls, remotePushUrls] = await Promise.all([
    runRequiredGitCommand(runner, registration, context, {
      operation: 'git.read_mirror_remote_names',
      arguments: ['--git-dir', mirrorPath, 'remote']
    }),
    runner.run(registration, context, {
      operation: 'git.read_mirror_remote_url',
      arguments: [
        '--git-dir',
        mirrorPath,
        'config',
        '--local',
        '--get-all',
        `remote.${remote.name}.url`
      ]
    }),
    runner.run(registration, context, {
      operation: 'git.read_mirror_remote_push_url',
      arguments: [
        '--git-dir',
        mirrorPath,
        'config',
        '--local',
        '--get-all',
        `remote.${remote.name}.pushurl`
      ]
    })
  ])
  const names = remoteNames.stdout.split(/\r?\n/).filter(Boolean)
  if (names.length !== 1 || names[0] !== remote.name) {
    throw new GitRuntimeError(
      'git.remote_mismatch',
      'Managed Git mirror contains unexpected named remotes.'
    )
  }
  if (remoteUrls.exitCode !== 0) {
    throw new GitRuntimeError(
      'git.remote_missing',
      'Managed Git mirror does not contain the selected remote.'
    )
  }
  const urls = remoteUrls.stdout.split(/\r?\n/).filter(Boolean)
  if (urls.length !== 1) {
    throw new GitRuntimeError(
      'git.remote_mismatch',
      'Managed Git mirror has an ambiguous remote URL.'
    )
  }
  if (remotePushUrls.exitCode === 0 && remotePushUrls.stdout.trim()) {
    throw new GitRuntimeError(
      'git.remote_mismatch',
      'Managed Git mirror must not configure a push-only remote URL.'
    )
  }
  if (remotePushUrls.exitCode !== 0 && remotePushUrls.exitCode !== 1) {
    requireGitCommandSuccess(remotePushUrls)
  }
  await assertNoMirrorConfigurationRewrite(runner, registration, context, mirrorPath)
  const observedUrl = urls[0]
  const observedSource = parseGitRemoteSource(observedUrl)
  assertGitRemoteIdentity(remote.source.identity, observedSource.identity)
}

async function fetchMirrorRemote(
  runner: GitCommandRunner,
  registration: GitExecutableRegistration,
  context: GitExecutionContext,
  mirrorPath: string,
  remoteName: string
): Promise<void> {
  assertGitRemoteName(remoteName)
  await runRequiredGitCommand(runner, registration, context, {
    operation: 'git.fetch_managed_mirror',
    arguments: [
      '--git-dir',
      mirrorPath,
      'fetch',
      '--no-write-fetch-head',
      '--prune',
      '--prune-tags',
      remoteName,
      '+refs/heads/*:refs/remotes/' + remoteName + '/*',
      '+refs/tags/*:refs/tags/*'
    ]
  })
}

async function assertNoMirrorConfigurationRewrite(
  runner: GitCommandRunner,
  registration: GitExecutableRegistration,
  context: GitExecutionContext,
  mirrorPath: string
): Promise<void> {
  for (const pattern of ['^url\\..*\\.insteadOf$', '^include\\.']) {
    const result = await runner.run(registration, context, {
      operation: 'git.inspect_mirror_rewrite_configuration',
      arguments: ['--git-dir', mirrorPath, 'config', '--local', '--get-regexp', pattern]
    })
    if (result.exitCode === 1) continue
    requireGitCommandSuccess(result)
    if (result.stdout.trim()) {
      throw new GitRuntimeError(
        'git.remote_invalid',
        'Managed Git mirror contains a local configuration rewrite or include.'
      )
    }
  }
}

function assertNamedRemoteIntegrity(remote: GitNamedRemote): void {
  assertGitNamedRemote(remote)
}

async function assertPathAbsent(
  path: string,
  code: 'git.mirror_exists' | 'git.operation_failed',
  message: string
): Promise<void> {
  try {
    await lstat(path)
  } catch (error) {
    if (isNodeCode(error, 'ENOENT')) return
    throw gitRuntimeFailure(
      'git.operation_failed',
      'Managed Git path could not be inspected.',
      error
    )
  }
  throw new GitRuntimeError(code, message)
}

async function assertDirectRegularBundleFile(bundlePath: string): Promise<void> {
  if (
    typeof bundlePath !== 'string' ||
    bundlePath.length === 0 ||
    bundlePath.includes('\u0000') ||
    !nodePath.isAbsolute(bundlePath) ||
    nodePath.normalize(bundlePath) !== bundlePath ||
    nodePath.parse(bundlePath).root === bundlePath
  ) {
    throw new GitRuntimeError('git.repository_invalid', 'Managed Git bundle path is invalid.')
  }
  let metadata: Awaited<ReturnType<typeof lstat>>
  try {
    metadata = await lstat(bundlePath)
  } catch (error) {
    throw gitRuntimeFailure('git.repository_invalid', 'Managed Git bundle is unavailable.', error)
  }
  if (metadata.isSymbolicLink() || !metadata.isFile()) {
    throw new GitRuntimeError(
      'git.repository_invalid',
      'Managed Git bundle must be a direct regular file.'
    )
  }
}

async function assertSameFilesystem(
  left: string,
  right: string,
  installationId: string
): Promise<void> {
  try {
    const [leftMetadata, rightMetadata] = await Promise.all([stat(left), stat(right)])
    if (leftMetadata.dev !== rightMetadata.dev) {
      throw new GitRuntimeError(
        'git.operation_failed',
        'Managed Git staging and mirror publication directories differ.',
        {
          installationId
        }
      )
    }
  } catch (error) {
    if (error instanceof GitRuntimeError) throw error
    throw gitRuntimeFailure(
      'git.operation_failed',
      'Managed Git publication directories could not be inspected.',
      error,
      {
        installationId
      }
    )
  }
}

function isNodeCode(value: unknown, expected: string): boolean {
  return Boolean(value && typeof value === 'object' && 'code' in value && value.code === expected)
}

File: networking/internal/gitcheckout/mirror.go
package gitcheckout

import (
	"context"
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

// Mirror is the readback identity of a verified launcher-owned bare mirror.
type Mirror struct {
	Path   string      `json:"path"`
	Remote NamedRemote `json:"remote"`
}

// rewritePatterns are the local configuration that could make a mirror fetch from
// somewhere other than the URL that was confirmed: a URL rewrite, and an include
// that could pull one in from a file outside the mirror.
var rewritePatterns = []string{`^url\..*\.insteadOf$`, `^include\.`}

// CreateMirror creates a new bare mirror through a private staging path and one
// atomic publication rename, so a reader never observes a half-built mirror.
func CreateMirror(ctx context.Context, runner *Runner, executable Executable, context ExecutionContext, paths InstallationPaths, remote NamedRemote) (Mirror, error) {
	return createMirror(ctx, runner, executable, context, paths, remote, "")
}

// CreateMirrorFromBundle imports one previously verified bundle into a fresh
// managed mirror without persisting the bundle as a git remote. Branches are
// written straight into the remote-tracking namespace, so resolving a revision
// afterwards is identical to a mirror created from the declared network remote.
func CreateMirrorFromBundle(ctx context.Context, runner *Runner, executable Executable, context ExecutionContext, paths InstallationPaths, remote NamedRemote, bundlePath string) (Mirror, error) {
	if err := assertDirectRegularBundleFile(bundlePath); err != nil {
		return Mirror{}, err
	}
	return createMirror(ctx, runner, executable, context, paths, remote, bundlePath)
}

func createMirror(ctx context.Context, runner *Runner, executable Executable, context ExecutionContext, paths InstallationPaths, remote NamedRemote, bundlePath string) (Mirror, error) {
	if err := AssertNamedRemote(remote); err != nil {
		return Mirror{}, err
	}
	lock, err := AcquireOperationLock(paths)
	if err != nil {
		return Mirror{}, err
	}
	defer func() {
		_ = lock.Release()
	}()
	if err = EnsureInstallationDirectories(paths); err != nil {
		return Mirror{}, err
	}
	if err = assertPathAbsent(paths.MirrorPath, ErrMirrorExists, "Managed Git mirror already exists."); err != nil {
		return Mirror{}, err
	}
	stagingID, err := newOwnerID()
	if err != nil {
		return Mirror{}, fmt.Errorf("%w: Managed Git staging mirror could not be named.", ErrOperationFailed)
	}
	stagingMirrorPath := filepath.Join(paths.StagingPath, "mirror-"+stagingID+".git")
	if err = assertPathAbsent(stagingMirrorPath, ErrOperationFailed, "Managed Git staging mirror already exists."); err != nil {
		return Mirror{}, err
	}
	if err = assertSameFilesystem(paths.StagingPath, filepath.Dir(paths.MirrorPath)); err != nil {
		return Mirror{}, err
	}
	if _, err = RunRequiredGitCommand(ctx, runner, executable, context, Command{
		Operation: "git.create_bare_mirror",
		Arguments: []string{"init", "--bare", stagingMirrorPath},
	}); err != nil {
		return Mirror{}, err
	}
	if bundlePath != "" {
		if _, err = RunRequiredGitCommand(ctx, runner, executable, context, Command{
			Operation: "git.import_bundle_objects",
			Arguments: []string{
				"--git-dir", stagingMirrorPath,
				"fetch", "--atomic", "--no-write-fetch-head", "--no-tags", "--",
				bundlePath,
				"+refs/heads/*:refs/remotes/" + remote.Name + "/*",
				"+refs/tags/*:refs/tags/*",
			},
		}); err != nil {
			return Mirror{}, err
		}
	}
	if _, err = RunRequiredGitCommand(ctx, runner, executable, context, Command{
		Operation: "git.add_mirror_remote",
		Arguments: []string{
			"--git-dir", stagingMirrorPath,
			"remote", "add", remote.Name, remote.Source.DeclaredURL,
		},
	}); err != nil {
		return Mirror{}, err
	}
	if bundlePath == "" {
		if err = fetchMirrorRemote(ctx, runner, executable, context, stagingMirrorPath, remote.Name); err != nil {
			return Mirror{}, err
		}
	}
	if err = VerifyBareMirror(ctx, runner, executable, context, stagingMirrorPath, remote); err != nil {
		return Mirror{}, err
	}
	if err = os.Rename(stagingMirrorPath, paths.MirrorPath); err != nil {
		return Mirror{}, fmt.Errorf("%w: Managed Git mirror could not be published atomically.", ErrOperationFailed)
	}
	if err = AssertManagedTarget(paths, paths.MirrorPath, "Managed Git mirror"); err != nil {
		return Mirror{}, err
	}
	if err = VerifyBareMirror(ctx, runner, executable, context, paths.MirrorPath, remote); err != nil {
		return Mirror{}, err
	}
	return Mirror{Path: paths.MirrorPath, Remote: remote}, nil
}

// FetchMirror fetches only the declared remote into an existing verified mirror.
func FetchMirror(ctx context.Context, runner *Runner, executable Executable, context ExecutionContext, paths InstallationPaths, remote NamedRemote) (Mirror, error) {
	if err := AssertNamedRemote(remote); err != nil {
		return Mirror{}, err
	}
	lock, err := AcquireOperationLock(paths)
	if err != nil {
		return Mirror{}, err
	}
	defer func() {
		_ = lock.Release()
	}()
	if err = AssertManagedTarget(paths, paths.MirrorPath, "Managed Git mirror"); err != nil {
		return Mirror{}, err
	}
	if err = VerifyBareMirror(ctx, runner, executable, context, paths.MirrorPath, remote); err != nil {
		return Mirror{}, err
	}
	// The fetch happens between two verifications: a mirror whose configuration was
	// changed after the first check would be caught by the second.
	if err = fetchMirrorRemote(ctx, runner, executable, context, paths.MirrorPath, remote.Name); err != nil {
		return Mirror{}, err
	}
	if err = VerifyBareMirror(ctx, runner, executable, context, paths.MirrorPath, remote); err != nil {
		return Mirror{}, err
	}
	return Mirror{Path: paths.MirrorPath, Remote: remote}, nil
}

// InspectMirror reads a managed mirror without fetching or changing anything.
func InspectMirror(ctx context.Context, runner *Runner, executable Executable, context ExecutionContext, paths InstallationPaths, remote NamedRemote) (Mirror, error) {
	if err := AssertNamedRemote(remote); err != nil {
		return Mirror{}, err
	}
	if err := AssertManagedTarget(paths, paths.MirrorPath, "Managed Git mirror"); err != nil {
		return Mirror{}, err
	}
	if err := VerifyBareMirror(ctx, runner, executable, context, paths.MirrorPath, remote); err != nil {
		return Mirror{}, err
	}
	return Mirror{Path: paths.MirrorPath, Remote: remote}, nil
}

// VerifyBareMirror verifies both the bare-repository state and the single remote
// identity the mirror is allowed to carry. A mirror is only usable if it is bare,
// has exactly the named remote, has one URL that is still the confirmed source,
// has no push-only URL, and carries no local configuration that could redirect it.
func VerifyBareMirror(ctx context.Context, runner *Runner, executable Executable, context ExecutionContext, mirrorPath string, remote NamedRemote) error {
	bare, err := RunRequiredGitCommand(ctx, runner, executable, context, Command{
		Operation: "git.verify_bare_mirror",
		Arguments: []string{"--git-dir", mirrorPath, "rev-parse", "--is-bare-repository"},
	})
	if err != nil {
		return err
	}
	bareState, err := RequireSingleGitLine(bare.Stdout, "Bare mirror state")
	if err != nil {
		return err
	}
	if bareState != "true" {
		return fmt.Errorf("%w: Managed Git mirror is not a bare repository.", ErrRepositoryNotBare)
	}
	remoteNames, err := RunRequiredGitCommand(ctx, runner, executable, context, Command{
		Operation: "git.read_mirror_remote_names",
		Arguments: []string{"--git-dir", mirrorPath, "remote"},
	})
	if err != nil {
		return err
	}
	if names := nonEmptyLines(remoteNames.Stdout); len(names) != 1 || names[0] != remote.Name {
		return fmt.Errorf("%w: Managed Git mirror contains unexpected named remotes.", ErrRemoteMismatch)
	}
	remoteURLs, err := runner.Run(ctx, executable, context, Command{
		Operation: "git.read_mirror_remote_url",
		Arguments: []string{"--git-dir", mirrorPath, "config", "--local", "--get-all", "remote." + remote.Name + ".url"},
	})
	if err != nil {
		return err
	}
	if remoteURLs.ExitCode == nil || *remoteURLs.ExitCode != 0 {
		return fmt.Errorf("%w: Managed Git mirror does not contain the selected remote.", ErrRemoteMissing)
	}
	urls := nonEmptyLines(remoteURLs.Stdout)
	if len(urls) != 1 {
		return fmt.Errorf("%w: Managed Git mirror has an ambiguous remote URL.", ErrRemoteMismatch)
	}
	remotePushURLs, err := runner.Run(ctx, executable, context, Command{
		Operation: "git.read_mirror_remote_push_url",
		Arguments: []string{"--git-dir", mirrorPath, "config", "--local", "--get-all", "remote." + remote.Name + ".pushurl"},
	})
	if err != nil {
		return err
	}
	pushExit := -1
	if remotePushURLs.ExitCode != nil {
		pushExit = *remotePushURLs.ExitCode
	}
	if pushExit == 0 && strings.TrimSpace(remotePushURLs.Stdout) != "" {
		return fmt.Errorf("%w: Managed Git mirror must not configure a push-only remote URL.", ErrRemoteMismatch)
	}
	if pushExit != 0 && pushExit != 1 {
		return RequireSuccess(remotePushURLs)
	}
	if err = assertNoMirrorConfigurationRewrite(ctx, runner, executable, context, mirrorPath); err != nil {
		return err
	}
	observedSource, err := ParseSource(urls[0])
	if err != nil {
		return err
	}
	return AssertIdentity(remote.Source.Identity, observedSource.Identity)
}

// assertNoMirrorConfigurationRewrite refuses a local URL rewrite or include: both
// would let the mirror fetch from an address other than the one that was
// confirmed, which is the one thing the mirror's identity exists to prevent.
func assertNoMirrorConfigurationRewrite(ctx context.Context, runner *Runner, executable Executable, context ExecutionContext, mirrorPath string) error {
	for _, pattern := range rewritePatterns {
		result, err := runner.Run(ctx, executable, context, Command{
			Operation: "git.inspect_mirror_rewrite_configuration",
			Arguments: []string{"--git-dir", mirrorPath, "config", "--local", "--get-regexp", pattern},
		})
		if err != nil {
			return err
		}
		if result.ExitCode != nil && *result.ExitCode == 1 {
			continue
		}
		if err = RequireSuccess(result); err != nil {
			return err
		}
		if strings.TrimSpace(result.Stdout) != "" {
			return fmt.Errorf("%w: Managed Git mirror contains a local configuration rewrite or include.", ErrRemoteInvalid)
		}
	}
	return nil
}

// fetchMirrorRemote fetches only the named remote's branches and tags into the
// remote-tracking namespace the launcher resolves revisions from.
func fetchMirrorRemote(ctx context.Context, runner *Runner, executable Executable, context ExecutionContext, mirrorPath, remoteName string) error {
	if err := AssertRemoteName(remoteName); err != nil {
		return err
	}
	_, err := RunRequiredGitCommand(ctx, runner, executable, context, Command{
		Operation: "git.fetch_managed_mirror",
		Arguments: []string{
			"--git-dir", mirrorPath,
			"fetch", "--no-write-fetch-head", "--prune", "--prune-tags",
			remoteName,
			"+refs/heads/*:refs/remotes/" + remoteName + "/*",
			"+refs/tags/*:refs/tags/*",
		},
	})
	return err
}

// assertPathAbsent refuses to build on top of something that is already there.
func assertPathAbsent(path string, code error, message string) error {
	if _, err := os.Lstat(path); err != nil {
		if os.IsNotExist(err) {
			return nil
		}
		return fmt.Errorf("%w: Managed Git path could not be inspected.", ErrOperationFailed)
	}
	return fmt.Errorf("%w: %s", code, message)
}

// assertDirectRegularBundleFile refuses anything but a canonical absolute path to
// a regular file that is not a link: a bundle is imported by path, and a link
// there would import something the caller never verified.
func assertDirectRegularBundleFile(bundlePath string) error {
	if err := AssertCanonicalAbsolutePath(bundlePath, "Managed Git bundle path", ErrRepositoryInvalid); err != nil {
		return err
	}
	metadata, err := os.Lstat(bundlePath)
	if err != nil {
		return fmt.Errorf("%w: Managed Git bundle is unavailable.", ErrRepositoryInvalid)
	}
	if metadata.Mode()&os.ModeSymlink != 0 || !metadata.Mode().IsRegular() {
		return fmt.Errorf("%w: Managed Git bundle must be a direct regular file.", ErrRepositoryInvalid)
	}
	return nil
}

// assertSameFilesystem keeps the publication rename inside one filesystem: a
// rename across devices is a copy, and a copy is not atomic.
func assertSameFilesystem(left, right string) error {
	leftDevice, err := deviceOf(left)
	if err != nil {
		return fmt.Errorf("%w: Managed Git publication directories could not be inspected.", ErrOperationFailed)
	}
	rightDevice, err := deviceOf(right)
	if err != nil {
		return fmt.Errorf("%w: Managed Git publication directories could not be inspected.", ErrOperationFailed)
	}
	if leftDevice != rightDevice {
		return fmt.Errorf("%w: Managed Git staging and mirror publication directories differ.", ErrOperationFailed)
	}
	return nil
}

func deviceOf(path string) (int64, error) {
	info, err := os.Stat(path)
	if err != nil {
		return 0, err
	}
	fingerprint, err := readFingerprint(path, info)
	if err != nil {
		return 0, err
	}
	return fingerprint.Device, nil
}

// nonEmptyLines lists the non-empty lines of a command's stdout.
func nonEmptyLines(stdout string) []string {
	lines := make([]string, 0, 1)
	for _, line := range strings.Split(strings.ReplaceAll(stdout, "\r\n", "\n"), "\n") {
		if line != "" {
			lines = append(lines, line)
		}
	}
	return lines
}
